### PR TITLE
Metric correlations

### PR DESCRIPTION
**Pull request by MrZammler**
##### Summary

This is a first attempt at implementing/copying [ks_2samp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html) which is used by cloud to run metric correlations. It follows the path for `two-sided` in mode `asymp`.

It will respond to api call at `/api/v1/metric_correlations` where it needs 4 inputs that describe 2 timeframes: baseline (defined by `baseline_after` and `baseline_before`) and a highlight area (defined by `highlight_after` and `highlight_before`) and will return a json similar to what is returned now in the cloud.

##### Test Plan

Still in testing, mostly to figure out if numbers match the python implementation, and for cloud to be able to consume the json response (differs slightly than cloud implementation in that it contains also chart contexts).

### CODE REVIEW

**Comment by andrewm4894** [comment]
this looking good to me - matches perfectly with prod results.

https://colab.research.google.com/drive/1PH0loJvWJFCgqD0vzbI2RzQdvzlfu1ck?usp=sharing

image
image
image

here is a link to a snapshot where i hit the endpoint 5 times. https://drive.google.com/file/d/1k5wZGY_CwmOver-9KxJBwHfB0IlJckrB/view?usp=sharing

some charts for overhead impact. 

**tldr** - looks like a 1sec jump to 30-50% of cpu on 1 core. 

image
image
image
image

**Comment by andrewm4894** [comment]
im going to put some cpu load on my dev node and see impact then.

**Comment by MrZammler** [comment]
I haven't checked generally the load on the machine, and will be the first thing to do. I will try to get some numbers, to see how it works on specific systems, with what range of data, etc.

Thanks Andy!!

**Comment by andrewm4894** [comment]
here is some rough stuff where i abused it a little. 

last big green was where i did a highlight window of 30 mins. other too middle greens were 10 min highlights. 

image

**Comment by andrewm4894** [comment]
@MrZammler i have tested the max_points and looks to be working now. 

Below shows impact of same query but for max points of 100, 1000, 500, 200, 100

image

**Comment by andrewm4894** [comment]
@MrZammler would you be able to add a `[metric_correlations]` section to `netdata.conf` as part of this PR and have it default to `enabled = yes`

**Comment by andrewm4894** [comment]
@juacker what would the BE need to be able to add `mc_capable` and `mc_enabled` fields to the `node_instance_capabilities ` table in the BE? 

I think we should lay whatever foundations we need for that as part of this PR too.

**Comment by juacker** [comment]
> @juacker what would the BE need to be able to add `mc_capable` and `mc_enabled` fields to the `node_instance_capabilities ` table in the BE?
> 
> I think we should lay whatever foundations we need for that as part of this PR too.

In order to identify this capability, the agent should list it in the `UpdateNodeInfo` aclk message sent to the cloud.

**Comment by MrZammler** [comment]
Config option added, will be default enabled for the tests (and default disabled for merge).

File: database/metric_correlations.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * computes the complementary cumulative probability P[D_n >= x] of the
+ * 2-sided 1-sample Kolmogorov-Smirnov distribution with sample size n at x.
+ * It returns at least 10 decimal digits of precision for n <= 500,
+ * at least 6 decimal digits of precision for 500 < n <= 200000,
+ * and a few correct decimal digits for n > 200000.
+ *
+ */
+
+double KSfbar (int n, double x);
```
Comment (vkalintiris) [inline database/metric_correlations.h:23]: Do we really need to make `KSfbar` available anywhere else?

File: database/metric_correlations.c [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        error("Metric correlations functionality is not enabled.");
+        buffer_strcat(wb, "{\"error\": \"Metric correlations functionality is not enabled.\" }");
+        return;
+    }
+
+    if (highlight_before <= highlight_after || baseline_before <= baseline_after) {
+        error("Invalid baseline or highlight ranges.");
+        buffer_strcat(wb, "{\"error\": \"Invalid baseline or highlight ranges.\" }");
+        return;
+    }
```
Comment (vkalintiris) [inline database/metric_correlations.c:225]: We also need to clamp `baseline_after`, `baseline_before`, `highlight_after`, `highlight_before`, to some `[min, max]` range and (a) either run metric correlation, or (b) return an appropriate error message.

File: database/metric_correlations.c [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    RRDSET *st;
+    size_t c = 0;
+    BUFFER *wdims = buffer_create(1000);
+
+    if (!max_points || max_points > MAX_POINTS)
+        max_points = MAX_POINTS;
+
+    //dont lock here and wait for results
+    //get the charts and run mc after
+    //should not be a problem for the query
```
Comment (vkalintiris) [inline database/metric_correlations.c:237]: To do this, we'd also need to set/unset an atomic flag that disabled deletion of charts.

**Review by vkalintiris — COMMENTED** [review]
Overall this LGTM. There's only one important comment regarding the way we are iterating charts.

I'd prefer us to maintain this under `ml/metric-corellation` but this is something I want to do myself later on, because it'd require moving `ml/*` stuff under a new subdir, eg. `ml/anomaly-detection.

File: database/metric_correlations.c (hunk 3 above)
Comment (MrZammler) [reply database/metric_correlations.c:237]: Yes! We discussed this already, thanks Vasili. I will look into it.

File: database/metric_correlations.h (hunk 1 above)
Comment (MrZammler) [reply database/metric_correlations.h]: No we don't! Thanks, removed!

File: database/metric_correlations.c (hunk 3 above)
Comment (MrZammler) [reply database/metric_correlations.c:237]: I've added locks with `rrdset_rdlock(st)`, should do it...

File: database/metric_correlations.c (hunk 2 above)
Comment (MrZammler) [reply database/metric_correlations.c:225]: Not sure yet what constitutes invalid range, aside from obvious being the before before after and such. Perhaps we will add, e.g. if we don't have any metrics during that period? But I guess since the input will come from the cloud which will be highlighted on a region containing metrics, it should be ok. We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes.

**Review by andrewm4894 — APPROVED** [review]

File: database/metric_correlations.c (hunk 3 above)
Comment (vkalintiris) [reply database/metric_correlations.c:237]: > I've added locks with rrdset_rdlock(st), should do it... ?

Not sure either... Can you run this under a thread sanitizer and hit the endpoint?

Comment (vkalintiris) [reply database/metric_correlations.c:237]: > I've added locks with rrdset_rdlock(st), should do it... ?

Not sure either... Can you run this under a thread sanitizer and hit the endpoint?

File: database/metric_correlations.c (hunk 2 above)
Comment (vkalintiris) [reply database/metric_correlations.c:225]: What I mean is that we should work with @andrewm4894, in order to come up with the largest duration we are willing to allow the agents to serve. IOW, we don't want to run metric correlation for a duration of 1 month (as this will most probably break the agent in some way).

> `error("Invalid baseline or highlight ranges.");`

This is a coalesced error message. It's not clear which invalid case we actually hit from the message description. We should return a separate one for an invalid `baseline` range and another one for the `highlight` range.

> We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes

Agreed, but if we don't have a predefined limit someone could DoS the agent.

Comment (vkalintiris) [reply database/metric_correlations.c:225]: What I mean is that we should work with @andrewm4894, in order to come up with the largest duration we are willing to allow the agents to serve. IOW, we don't want to run metric correlation for a duration of 1 month (as this will most probably break the agent in some way).

> `error("Invalid baseline or highlight ranges.");`

This is a coalesced error message. It's not clear which invalid case we actually hit from the message description. We should return a separate one for an invalid `baseline` range and another one for the `highlight` range.

> We will see, I'm expecting the cloud to kinda safe-guard the values, but we will protect if we want, yes

Agreed, but if we don't have a predefined limit someone could DoS the agent.

**Review by iigorkarpov — DISMISSED** [review]
LGTM

**Comment by thiagoftsm** [comment]
Hello @MrZammler ,

Please, take a look in [LGTM](https://github.com/netdata/netdata/pull/12582#issuecomment-1111994871)!

File: database/metric_correlations.c (hunk 3 above)
Comment (MrZammler) [reply database/metric_correlations.c:237]: It should be ok... I've tried it, (after reducing number of charts to ~50) but we could discuss it from Monday maybe. We could also add setting `st->last_accessed_time` to keep it from getting obsolete for a while.

File: database/metric_correlations.c (hunk 2 above)
Comment (MrZammler) [reply database/metric_correlations.c:225]: Indeed. That part needs attention, and will be done right after. As is with the max_points, we will define a good number to balance performance and accuracy, but that will come I guess from testing it in real world.

**Review by vkalintiris — APPROVED** [review]

**Review by vkalintiris — APPROVED** [review]
@andrewm4894 lgtm, I think we can merge this.

**Review by stelfrag — APPROVED** [review]

**Comment by MrZammler** [comment]
Thanks all!
